File: DevEnvironment/DataStructures/src/main/java/datastructures/cache/core/CacheProperties.java
package datastructures.cache.core;

import base.core.Conditions;
import datastructures.cache.interfaces.ICacheProperties;

/**
 * The CacheProperties class implements properties of a cache.
 */
public final class CacheProperties implements ICacheProperties {
    private final int capacity;
    private final int numberOfItemsForMakingAvailableSpace;

    /**
     * The CacheProperties constructor.
     */
    public CacheProperties(
        int capacity,
        int numberOfItemsForMakingAvailableSpace) {

        Conditions.validate(
            capacity > 0,
            "The capacity of a cache has to be positive.");

        Conditions.validate(
            numberOfItemsForMakingAvailableSpace <= capacity,
            "The number of items for making available space can not exceed the cache capacity.");

        this.capacity = capacity;
        this.numberOfItemsForMakingAvailableSpace = numberOfItemsForMakingAvailableSpace;
    }

    /**
     * Gets a capacity of a cache.
     */
    @Override
    public int getCapacity() {
        return this.capacity;
    }

    /**
     * Gets the number of items for making available space in the cache.
     */
    @Override
    public int getNumberOfItemsForMakingAvailableSpace() {
        return this.numberOfItemsForMakingAvailableSpace;
    }
}

File: DevEnvironmentCLion/DataStructures/cache/core/CacheProperties.cpp
#include "PreCompiled.h"
#include "CacheProperties.h"
#include "CacheException.h"

using namespace datastructures::cache;

/**
 * Creates properties of a cache.
 */
ICachePropertiesSharedPtr CacheProperties::make(
    size_t capacity,
    size_t numberOfItemsForMakingAvailableSpace) {

    return std::make_shared<CacheProperties>(capacity, numberOfItemsForMakingAvailableSpace);
}

/**
 * The CacheProperties constructor.
 */
CacheProperties::CacheProperties(
    size_t capacity,
    size_t numberOfItemsForMakingAvailableSpace) :
        m_capacity(capacity),
        m_numberOfItemsForMakingAvailableSpace(numberOfItemsForMakingAvailableSpace)
{
    if (numberOfItemsForMakingAvailableSpace > capacity)
    {
        std::string errorMessage = "The number of items for making available space can not exceed the cache capacity.";
        throw CacheException(errorMessage);
    }
}

/**
 * The CacheProperties destructor.
 */
CacheProperties::~CacheProperties()
{
}

/**
 * Gets a capacity of a cache.
 */
size_t CacheProperties::getCapacity() const
{
    return m_capacity;
}

/**
 * Gets the number of items for making available space in the cache.
 */
size_t CacheProperties::getNumberOfItemsForMakingAvailableSpace() const
{
    return m_numberOfItemsForMakingAvailableSpace;
}
